server/data_common/driver.py

```diff
@@ -0,0 +1,319 @@
+                    key_idx = (anno_data <= max_).ravel()
+                    mask = np.logical_and(mask, key_idx)
+
+        return mask
+
+    def _filter_to_mask(self, filter):
+        """
+        Return the filter as a row and column selection list.
+        No filter on a dimension means 'all'
+        """
+        shape = self.get_X_array_shape()
+        var_selector = None
+        obs_selector = None
+        if filter is not None:
+            if Axis.OBS in filter:
+                obs_selector = self._axis_filter_to_mask(Axis.OBS, filter['obs'], shape[0])
+
+            if Axis.VAR in filter:
+                var_selector = self._axis_filter_to_mask(Axis.VAR, filter['var'], shape[1])
+
+        return (obs_selector, var_selector)
+
+    def check_new_labels(self, labels_df):
+        """Check the new annotations labels, then set the labels_df index"""
+        if labels_df is None or labels_df.empty:
+            return
+
+        labels_df.index = self.get_obs_index()
+
+        # all lables must have a name, which must be unique and not used in obs column names
+        if not labels_df.columns.is_unique:
+            raise KeyError(f"All column names specified in user annotations must be unique.")
+
+        # the label index must be unique, and must have same values the anndata obs index
+        if not labels_df.index.is_unique:
+            raise KeyError(f"All row index values specified in user annotations must be unique.")
+
+        obs_columns = self.get_obs_columns()
+
+        duplicate_columns = list(set(labels_df.columns) & set(obs_columns))
+        if len(duplicate_columns) > 0:
+            raise KeyError(
+                f"Labels file may not contain column names which overlap " f"with h5ad obs columns {duplicate_columns}"
+            )
+
+        # labels must have same count as obs annotations
+        obs_shape = self.get_obs_shape()
+        if labels_df.shape[0] != obs_shape[0]:
+            raise ValueError("Labels file must have same number of rows as data file.")
+
+    def data_frame_to_fbs_matrix(self, filter, axis):
+        """
+        Retrieves data 'X' and returns in a flatbuffer Matrix.
+        :param filter: filter: dictionary with filter params
+        :param axis: string obs or var
+        :return: flatbuffer Matrix
+
+        Caveats:
+        * currently only supports access on VAR axis
+        * currently only supports filtering on VAR axis
+        """
+
+        if axis != Axis.VAR:
+            raise ValueError("Only VAR dimension access is supported")
+
+        try:
+            obs_selector, var_selector = self._filter_to_mask(filter)
+        except (KeyError, IndexError, TypeError, AttributeError) as e:
+            raise FilterError(f"Error parsing filter: {e}") from e
+
+        if obs_selector is not None:
+            raise FilterError("filtering on obs unsupported")
+
+        X = self.get_X_array(obs_selector, var_selector)
+        X = MatrixProxy.create(X)
```

Comment:
why not just have get_X_array() return a MatrixProxy (or more accurately, a matrix implementing the proxy protocol)?

That would let data drivers / engines use the MatrixProxy class only when it is needed, and dispense with it when not.

Reply:
I'm not sure what the best solutions is.   If I put the MatrixProxy.create in the get_X_array, then I run into a problem.
The diffexp code uses get_X_array, and it is not expecting a MatrixProxy.  There is one case in the tests suite where it fails because it attempts to call "mean" and a few other function on the matrix proxy, which is not implemented.  

So the solution here is to create a proxy just before the encode_matrix_fbs, which is what the original code did.   For the tiledb case this is essentially a no-op.

Reply:
Good point.  As we discussed offline, the diffexp implementation currently has knowledge of which underlying data "adapter" or engine it is using.  Ie, diffexp is quite AnnData specific.   Even if we could generalize it, I suspect we will need adapter-specific implementations to achieve acceptable performance due to wide variations on assumptions about partitioning, etc.